sdk/identity/Azure.Identity/tests/Mock/MockIdentityClient.cs

```diff
@@ -10,7 +10,90 @@
+        {
+            return CreateToken(scopes, clientId: clientId, cancellationToken: cancellationToken);
+        }
+
+        public async override Task<AccessToken> AuthenticateAsync(string[] scopes, string clientId = null, CancellationToken cancellationToken = default)
+        {
+
+            return await CreateTokenAsync(scopes, clientId: clientId, cancellationToken: cancellationToken);
+        }
+
+        private async Task<AccessToken> CreateTokenAsync(string[] scopes, string tenantId = default, string clientId = default, string clientSecret = default, CancellationToken cancellationToken = default)
+        {
+            if (cancellationToken != default)
+            {
+                await Task.Delay(1000, cancellationToken);
+            }
+
+            return _tokenFactory(scopes, tenantId, clientId, clientSecret, cancellationToken);
+        }
+
+        private static AccessToken ExpiredTokenFactory(string[] scopes, string tenantId, string clientId, string clientSecret, CancellationToken cancellationToken)
+        {
+            return CreateAccessToken(scopes, tenantId, clientId, clientSecret, DateTimeOffset.UtcNow - TimeSpan.FromMinutes(1));
+        }
+
+        private static AccessToken LiveTokenFactory(string[] scopes, string tenantId, string clientId, string clientSecret, CancellationToken cancellationToken)
+        {
+            return CreateAccessToken(scopes, tenantId, clientId, clientSecret, DateTimeOffset.UtcNow + TimeSpan.FromHours(1));
+        }
+
+        private static AccessToken CreateAccessToken(string[] scopes, string tenantId, string clientId, string clientSecret, DateTimeOffset expires)
+        {
+            MockToken token = CreateMockToken(scopes, tenantId, clientId, clientSecret);
+
+            return new AccessToken(token.ToString(), expires);
+        }
+
+        private static MockToken CreateMockToken(string[] scopes, string tenantId, string clientId, string clientSecret)
+        {
+            return new MockToken().WithField("scopes", string.Join("+", scopes)).WithField("tenantId", tenantId).WithField("clientId", clientId).WithField("clientSecret", clientSecret);
+        }
+
+        public static MockManagedIdentityClient ExpiredTokenClient { get; } = new MockManagedIdentityClient(ExpiredTokenFactory);
+
+        public static MockManagedIdentityClient LiveTokenClient { get; } = new MockManagedIdentityClient(LiveTokenFactory);
+
+        private Func<string[], string, string, string, CancellationToken, AccessToken> _tokenFactory;
+
+        private AccessToken CreateToken(string[] scopes, string tenantId = default, string clientId = default, string clientSecret = default, CancellationToken cancellationToken = default)
+        {
+            if (cancellationToken != default)
+            {
+                Task.Delay(1000, cancellationToken).GetAwaiter().GetResult();
```

Comment:
@stephentoub, is this reliable? BTW, we are doing it as Thread.Sleep does not have an overload that takes a cancellation token.

Reply:
It's "reliable" in that it'll sleep for at least a second.  It is effectively sync-over-async, though, in that this will block the current thread until the async operation completes, which necessitates a thread pool thread briefly, so if this were production code and this was running on a thread pool thread, I'd be more concerned.

Thread.Sleep doesn't take a cancellation token as it just maps to the OS' thread sleep functionality, e.g. `::Sleep(milliseconds)` on Windows, where there's no notion of cancellation.

If this were production code, and you wanted to synchronously block the current thread for X milliseconds or until a cancellation request occurred, I'd probably instead advocate for something more like:
```C#
private static readonly ManualResetEventSlim s_sleeper = new ManualResetEventSlim();
...
s_sleeper.Wait(1000, cancellationToken);
```
but for test code, what you have is fine.